Improve directory structure and autosave feature

JasperSurmont opened the pull request:
This commit changes the directory structure as mentioned in #9. It shouldn't mess up any existing files, so users of the plugin should not directly be affected (except from getting used to the new structure).  

The main reason that I'd like this structure is that using the autosave you can have a timeline of the different heatmaps throughout your career. However, since files are not overwritten anymore, they will take up more space the more saves you do (especially the images). Maybe an option could be made that sets a limit to the amount of images?

The proposed structure is as follows:

```
worldheatmap/
├─ Heatmap Files/
│  ├─ [userId1]/
│  │  ├─ 240617-2042.heatmaps
│  ├─ [userId2]/
│  │  ├─ 240611-2010.heatmaps
├─ Heatmap Images/
│  ├─ [userId1]/
│  │  ├─ 240616-2212TYPE_A.tif
│  │  ├─ 240616-2212TYPE_B.tif
│  ├─ [userId2]/
│  │  ├─ 240616-2100TYPE_A.tif
```

It now also saves explicit saves and autosaves in the same folders, as it will not overwrite anything anymore.

Commit:
Fix autosave feature and improve directory structure

Commit:
Revert Gradle property change

JasperSurmont commented:
Also, the naming of the files is now abstracted to another class. This makes it less prone to small errors when you need to craft the filename every time separately.

JasperSurmont commented on src/main/java/com/worldheatmap/WorldHeatmapPlugin.java:
```diff
@@ -411,10 +410,6 @@ public void onNpcLootReceived(final NpcLootReceived npcLootReceived) {
-        File heatmapsFile = new File(HEATMAP_FILES_DIR, mostRecentLocalUserID + ".heatmaps");
-        File typeAImageFile = new File(HEATMAP_IMAGE_DIR, mostRecentLocalUserID + "_" + HeatmapNew.HeatmapType.TYPE_A + ".tif");
-        File typeBImageFile = new File(HEATMAP_IMAGE_DIR, mostRecentLocalUserID + "_" + HeatmapNew.HeatmapType.TYPE_B + ".tif");
-
```
Not necessary to create the files when it's not autosave anyway. I moved it to inside the if / else

GrandTheftWalrus commented:
Sweet, I'll check this out either tomorrow after class or on Friday. 

Also, I just remembered something upon seeing this pull request which I should've mentioned earlier. I only took a cursory gander at the code (which looks good) so I'm not sure if you've already addressed this, but: If the location at which the plugin saves and loads heatmap files is changed, then the save files of existing users won't be found unless there's a check performed for the old save location (which should move/remove them from the old location too). Abstracting the heatmap file logic to its own class makes this pretty easy though, so good idear.

JasperSurmont commented:
I think I saw this, because it used to be stored in a different format?

I left that there for the time being.

GrandTheftWalrus commented on src/main/java/com/worldheatmap/WorldHeatmapPlugin.java:
```diff
@@ -114,15 +112,16 @@ WorldHeatmapConfig provideConfig(ConfigManager configManager) {
+        log.debug("Loading most recent heatmaps under user ID {}...", mostRecentLocalUserID);
 
+
+        File filepathUserID = HeatmapFile.getLastHeatmap(mostRecentLocalUserID);
```
A check should be made inside HeatmapFile.getLastHeatmap() to see if the file `Heatmap Files\userId.heatmaps` exists, because this is where .heatmaps files are currently saved to/loaded from, so existing users will have their progress saved here. If it does exist, then move the file to its new proper location in `Heatmap Files\userID\yyyy-mm-dd.heatmaps` before returning it as a File

JasperSurmont replied:
Correct, I'll do it tonight

GrandTheftWalrus reviewed (changes requested):
I just made le one comment and haven't done a full review yet

Commit:
Add transition from old version

JasperSurmont commented:
Should be implemented and it worked on my (Linux) machine. Should probably be tested on Windows / Mac too

GrandTheftWalrus committed:
Merging in PR to change logging from 'info' level to 'debug'

GrandTheftWalrus committed:
Refactored the ugly method that handles config changes enabling/disabling heatmaps

GrandTheftWalrus committed:
Refactored another ugly method that checks which heatmap types are enabled (by reading the config methods)

GrandTheftWalrus committed:
Addressed a bug where the check for legacy heatmap locations would assume that the legacy file exists and try to move it, causing an exception if it doesn't (e.g. upon fresh install when neither the legacy or new files exist). Also changed .heatmaps file date format to yyyy-MM-dd_hh-mm and made it so that getMostRecentFile() ignores files without parseable dates when looking for the latest file

GrandTheftWalrus committed:
Moved method WorldHeatmapPlugin.readHeatmapOld() to HeatmapNew.readLegacyHeatmapFile()

GrandTheftWalrus committed:
-Made it so that a new dated file is only created upon backup routine or heatmap type reset, rather than every time a .heatmaps file is saved.
-Fixed how GameTimeTicks wasn't actually being recorded upon reading a .heatmaps file (this is probably why the original backup feature wasn't working. It would only make a backup after 6 hours of gameplay by default, but this timer would reset on login/logout so it wasn't possible)

GrandTheftWalrus committed:
-Refactored handling of legacy heatmap files, and made it so that legacy V1 (the extra old type) heatmap files are renamed to .old, and if a newer heatmap doesn't exist then it converts, loads, and saves the legacy V1 type as the new type.
-Renames old "Backups" folder to "Backups.old"
-Random minor fixes/refactoring/removing unused code

GrandTheftWalrus commented:
@JasperSurmont Did you see the PR that I made to your fork? I would add ye as a reviewer but can't

JasperSurmont committed:
Merge pull request #1 from GrandTheftWalrus/bugfixing/new-directory-structure

Bugfixing/new directory structure

JasperSurmont commented:
@GrandTheftWalrus Sorry, I completely missed that. I'm not too familiar with PRs

JasperSurmont commented:
Is this ready to be merged or am I missing something? @GrandTheftWalrus

GrandTheftWalrus commented:
@JasperSurmont
Oh yeah, it's good. I'm going to do some more thorough testing before I make the PR to the RuneLite plugin hub, but I forgot I could approve this PR meanwhile (now that my PR to your fork went through). lots of PR-ing.

JasperSurmont commented:
@GrandTheftWalrus  Okay perfect, thanks. Could you let me know when the PR to RL Hub is approved?

GrandTheftWalrus commented:
> @GrandTheftWalrus Okay perfect, thanks. Could you let me know when the PR to RL Hub is approved?

Sure cuh 💯🅱
